### DISCUSSION TITLE
How do I use plugins

### DISCUSSION THREAD

**Discussion by marutypes** [swc-project/swc#2113] [Q&A] [answered]
Hi! 

At Shopify we're trying out implementing SWC in our jest tests (and likely later our build process) as an optimization over babel. Unfortunately, there are some home grown babel plugins we use that we will need to replace.

I noticed there is a [plugin section in the docs ](https://swc.rs/docs/usage-plugin) about how to make custom transforms, which I was hoping I could take advantage of. Unfortunately, I can't actually find an example of how to use these plugins (whether in the `.swcrc` or in the inline config for jest). 

I'd appreciate any guidance on this, or a link if I just missed a doc page 😅

Thanks!

**edit** Trawling through github I've found some references to [a `plugin` field on the config object when using SWC from js](https://github.com/chialab/rna/blob/main/packages/esbuild-plugin-swc/lib/index.js#L93), but none in an `.swcrc` file. I'm not sure how one would use this in `.swcrc` but I will try it out in inlined jest config for now and see if it works there. Any guidance would still be super appreciated, and I'd be down to PR the docs with it later if that is something I can do to help.

**Comment by kdy1** [comment]
Currently, there's no good plugin system. 
I'm going to provide a rust plugin API, but it still requires some decisions.

I think the js plugin is not a good fit for swc, and I'm going to remove js plugin api in v2. 
(I added it, but I faced some restrictions and I think I should remove it.)



Currently, many users depend on swc from rust side and build their own binaries, including wasm target. From rust-side PoV, swc is quite modular and I can help you if you are going to do so.

**Reply by marutypes** [reply]
> Currently, there's no good plugin system.

:(

> Currently, many users depend on swc from rust side and build their own binaries, including wasm target. From rust-side PoV, swc is quite modular and I can help you if you are going to do so.

We might be interested in trying this out! I'll DM you soon.

**Reply by kwonoj** [reply]
I'm on the exact same boat at our work, in half-way migration custom babel plugin for the jest and put things on hold.

Curious about for the further designs

- removing js-side plugin api sounds like blocking easier migration paths as it's enforced to write rustlang based plugins to port (and our restricted infra requires amount to setup rust build pipeline in CI even). Is this firm decision, or something preliminary interface can remain and let consumer opt in to workaround as they want?

- what'll be looks like if rust side plugin system arrives? any rough plans? If it's possible, would like to avoid build-own-binary approach for the poc / long term approaches.

**Reply by kdy1** [reply]
> removing js-side plugin api sounds like blocking easier migration paths as it's enforced to write rustlang based plugins to port (and our restricted infra requires amount to setup rust build pipeline in CI even). Is this firm decision, or something preliminary interface can remain and let consumer opt in to workaround as they want?

Passing ast to js to process it with js is very slow. It's why I'm going to remove it.


> what'll be looks like if rust side plugin system arrives? any rough plans? If it's possible, would like to avoid build-own-binary approach for the poc / long term approaches.

I'm not sure about this at the moment, but I'm considering publishing plugins as dynamic libraries and loading those libraries from `@swc/core`.

**Reply by kwonoj** [reply]
> Passing ast to js to process it with js is very slow.

I agree it's quite cost expensive and not able to be considered as long term (or recommended approach). Q is for the practical migration paths if someone have established pipeline based on existing tools like babel /or anything else - requiring to port / write rust plugin is stiff blocker to explore plugging in swc into existing pipeline. Would it be considerable to mark with disclaimer, or even mark as deprecated path but allow to opt in?

**Reply by kdy1** [reply]
The support for babel plugins is not implemented, so it can't be a deprecated path.

**Reply by kwonoj** [reply]
Double clarification, I'm saying paths for writing plugin in js, not directly using babel plugin itself. Writing / porting existing plugin into js vs. Rust is stiff differences for migrating / exploration.

**Reply by kdy1** [reply]
Ah, I see.
But the single js plugin means all ast should be serialized and passed to the js world.
I'm not sure if it will worth it. I'm okay with maintaining js API, but users have to write plugins in rust anyway.

**Reply by kwonoj** [reply]
Yes, as wrote I agree eventually plugins need to be written in rust. But even with rough edges if js api is alive, we can try paths like write quick, dirty, slow js plugin to verify migration poc works -> then convert those into rust one by one. Otherwise to prove swc works all the plugins have to be written in rust upfront.

**Reply by kdy1** [reply]
Ah, I see. It seems reasonable in some aspects, but still, I'm not sure about it.

https://github.com/vercel/next.js/tree/706547e0c3cdfcaf0e7c744aba9406642918aeb7/packages/next/build/swc

For next.js, we ported one plugin at a time, and we tested those transforms using a test system of `swc_ecma_transforms_testing`. 


Note: Testing the transforms from js is not a good idea, as many js testing tools assume full control over tty console.
Actually, such behavior of `jest` delayed tasks a lot.

**Reply by kdy1** [reply]
@kwonoj I think about this more.

I have a question regarding js api.



At now, is implementing a swc plugin in js easier than implementing it as a rust transform? If so, I'll preserve js plugin api, to allow user porting one plugin at a time. 
I thought it would be easier to implement a transform in rust because there are lots of helpers, but I'm too familiar with rust and swc to make fair decision.

**Reply by kwonoj** [reply]
It's totally personal opinion, but in my opinion it is distinct benefit to allow to write js based plugins for the existing js ecosystem:

- language differences: Not all of js devs are familiar with rustlang & its tooling, writing a plugin with new language itself is high barrier. It is easier with helpers & tools provided by swc itself but language difference itself matters.
- tooling, settings up, deployments.. : Not all of devs have flexible 3rd party CI system / or depends on local system can compile arbitary rust code & execute. If someone on the js land trying to use swc, it is safe to assume they have existing setups to run js codebases but compiling rust plugin is a different story.

For the `implementation` itself I honestly don't think writing js plugin is notably easier if we don't care about external differences (i.e language). Sometimes familiarity beats though.

**Reply by kdy1** [reply]
Thanks! I actually need personal opinions from someone who worked on rust transforms :)
I was curious about it because js plugin is almost bare metal.


It would be great if swc supports babel plugin as a fallback (so people migrate one babel plugin at a time), but it would require lots of work.

**Reply by kwonoj** [reply]
Yes, I noticed that too js plugin support is nearly `on your own`. If I can share my personal 5c, that'd be enough as long as core swc accept js side manipulations. It'd be better if existing plugin from babel works out of the box - but I'd consider that's more of userland implementation, not the responsibility of the core swc as it exponentially grows scope of the project.

**Reply by marutypes** [reply]
> At now, is implementing a swc plugin in js easier than implementing it as a rust transform? If so, I'll preserve js plugin api, to allow user porting one plugin at a time.
I thought it would be easier to implement a transform in rust because there are lots of helpers, but I'm too familiar with rust and swc to make fair decision.

As someone far more familiar with JS/TS and it's ecosystem than rustlang, I was able to get a basic plugin working in JS in a couple hours but I've most of this week trying to level up my rust skills to be able to get the custom binary path working at all.

I'm still interested in going down that path since performance is our main motivation, but its definitely way harder for someone used to the JS ecosystem.

(The CI / distribution is also much more daunting for an org/dev that has established patterns for JS)

**Reply by btoaelbore** [reply]
@kdy1 in our in-house build we are using swc plugin in javascript its easy for me because im most familiar with javascript

**Reply by joshcartme** [reply]
> https://github.com/vercel/next.js/tree/706547e0c3cdfcaf0e7c744aba9406642918aeb7/packages/next/build/swc
> 
> For next.js, we ported one plugin at a time, and we tested those transforms using a test system of `swc_ecma_transforms_testing`.

@kdy1 assuming I've created plugins the way it's done with next.js above, how do I use that with `swc-loader`.  `swc-loader` uses `swc/core`s `transform` and `transformSync` but I'd of course want to use my own.  Do I create my own loader as well?

Thanks!

**Comment by kwonoj** [comment, marked as answer]
It took long, but let me mark this as answered via https://github.com/swc-project/swc/discussions/3540 and would like to suggest to try out new plugin interfaces. There are rough edges still, hope to get some feedback to improve it.
